assets/js/admin/blocks-toggling-control.js

```diff
@@ -57,55 +57,59 @@ export const startBlocksTogglingControl = ( postType ) => {
 		return;
 	}
 
-	let initialSenseiBlocksCount;
-	let previousSenseiBlocksCount;
+	let initialWithSenseiBlocks; // Whether initial state has Sensei Blocks.
+	let previousWithSenseiBlocks; // Whether previous state has Sensei Blocks.
 	let lastBlocks;
 
 	editorLifecycle( {
 		subscribeListener: () => {
-			const newBlocks = select( 'core/editor' ).getBlocks();
+			const newBlocks = coreEditorSelector.getBlocks();
 
 			// Check if blocks were changed.
 			if ( newBlocks !== lastBlocks ) {
 				lastBlocks = newBlocks;
 				toggleLegacyMetaboxes();
 
-				previousSenseiBlocksCount = getSenseiBlocksCount();
+				previousWithSenseiBlocks = hasSenseiBlocks();
 
-				if ( undefined !== initialSenseiBlocksCount ) {
+				if ( undefined !== initialWithSenseiBlocks ) {
 					toggleLegacyOrBlocksNotice();
 				}
 			}
 		},
 		onSetDirty: () => {
-			// Set initial blocks count.
+			// Set initial blocks state.
 			if (
 				coreEditorSelector.isEditedPostDirty() &&
-				undefined === initialSenseiBlocksCount
+				undefined === initialWithSenseiBlocks
 			) {
-				initialSenseiBlocksCount = previousSenseiBlocksCount;
+				// If it will fill the template (needs_template is true),
+				// we consider that it has Sensei blocks initially.
+				initialWithSenseiBlocks =
+					coreEditorSelector.getCurrentPostAttribute( 'meta' )
```

Comment:
Here's the main change of the PR.

The rest is only a refactor to use a boolean instead of a counter to check if it has or not the blocks.